### PR TITLE
Fix issue with operator OOM restart

### LINKED ISSUE #601: [Bug][Operator][Leader election?] Operator failure and restart, logs attached
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/kuberay/issues) and found no similar issues.

### KubeRay Component

ray-operator

### What happened + What you expected to happen

I was running the nightly KubeRay operator for development purposes and observed a failure with some strange error messages logged.

The logs had roughly 1000 lines of `Read request instance not found error` with names of pods unrelated to KubeRay,
followed by some complaints about leases and leader election,
followed by a crash and restart.

Ideally, there shouldn't be any issues related to leader election, since we [don't support leader election](https://github.com/ray-project/kuberay/issues/474) right now.

Only the last few lines of `Read request instance not found error!` are pasted below.

```
2022-09-28T18:52:13.868Z	INFO	controllers.RayCluster	Read request instance not found error!	{"name": "kube-system/konnectivity-agent-795c8dddf.171917992063a820"}
2022-09-28T18:52:13.869Z	INFO	controllers.RayCluster	Read request instance not found error!	{"name": "kube-system/konnectivity-agent-795c8dddf.171917a01bf956aa"}
2022-09-28T18:52:13.869Z	INFO	controllers.RayCluster	Read request instance not found error!	{"name": "kube-system/konnectivity-agent.171913fcc9e51685"}
2022-09-28T18:52:13.869Z	INFO	controllers.RayCluster	Read request instance not found error!	{"name": "kube-system/konnectivity-agent.1719140d14d73e54"}
2022-09-28T18:52:23.872Z	INFO	controllers.RayCluster	Read request instance not found error!	{"name": "kube-system/konnectivity-agent-795c8dddf-cqw2g.1719179920d0d07b"}
2022-09-28T18:52:53.884Z	INFO	controllers.RayCluster	Read request instance not found error!	{"name": "kube-system/konnectivity-agent-795c8dddf-pp5jv.171917a01c50dde1"}
2022-09-28T19:59:07.446Z	INFO	controllers.RayCluster	Read request instance not found error!	{"name": "default/gk3-autopilot-cluster-1-nap-1l3c0g4h-554d0488-ltwl.1718f0a38651b857"}
2022-09-28T19:59:07.447Z	INFO	controllers.RayCluster	Read request instance not found error!	{"name": "default/gk3-autopilot-cluster-1-nap-1l3c0g4h-83672b81-t6rs.1718f0a3d8493393"}
E0928 21:27:20.873442       1 leaderelection.go:330] error retrieving resource lock ray-system/ray-operator-leader: Get "https://10.8.56.1:443/api/v1/namespaces/ray-system/configmaps/ray-operator-leader": context deadline exceeded
2022-09-28T21:27:20.874Z	DEBUG	events	Normal	{"object": {"kind":"ConfigMap","apiVersion":"v1"}, "reason": "LeaderElection", "message": "kuberay-operator-56749d657d-x5f65_085aa91b-20ef-4011-b7f4-4edff573f5c7 stopped leading"}
2022-09-28T21:27:20.874Z	DEBUG	events	Normal	{"object": {"kind":"Lease","namespace":"ray-system","name":"ray-operator-leader","uid":"e6a67c4c-c3a9-4b83-a0c6-7a9a3f02a1a2","apiVersion":"coordination.k8s.io/v1","resourceVersion":"31046663"}, "reason": "LeaderElection", "message": "kuberay-operator-56749d657d-x5f65_085aa91b-20ef-4011-b7f4-4edff573f5c7 stopped leading"}
I0928 21:27:20.874714       1 leaderelection.go:283] failed to renew lease ray-system/ray-operator-leader: timed out waiting for the condition
2022-09-28T21:27:20.903Z	INFO	Stopping and waiting for non leader election runnables
2022-09-28T21:27:20.903Z	INFO	Stopping and waiting for leader election runnables
2022-09-28T21:27:20.903Z	INFO	Stopping and waiting for caches
2022-09-28T21:27:20.903Z	INFO	Stopping and waiting for webhooks
2022-09-28T21:27:20.903Z	INFO	Wait completed, proceeding to shutdown the manager
2022-09-28T21:27:20.903Z	ERROR	setup	problem running manager	{"error": "leader election lost"}
```

### Reproduction script

I don't know how to reproduce this yet.

### Anything else

I don't know yet.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by wilsonwang371**
## Why are these changes needed?

This is the patch to fix #601 issue.

In our kuberay controller, we are watching all the events so that we can find out if there is any failed health probe event. In case we see this event happening, we will try to create a new pod to replace the failed one currently. Ideally, we should not receive a large amount of these events. However, we are not correctly filtering out these messages. When a lot of unrelated events have arrived, we will end up printing too many log messages and make the controller run slow.

In this patch, we use filters to block the event messages that we do not care about as early as possible.

## Related issue number

#601 

## Checks

- [X] I've made sure the tests are passing. 
- Testing Strategy
   - [X] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 1]
```diff
@@ -121,27 +123,42 @@ func (r *RayClusterReconciler) eventReconcile(request ctrl.Request, event *corev
 	// we only care about pod events
 	if event.InvolvedObject.Kind != "Pod" || event.Type != "Warning" || event.Reason != "Unhealthy" ||
 		!strings.Contains(event.Message, "Readiness probe failed") {
+		// This is not supposed to happen
+		r.Log.Error(fmt.Errorf("unexpected event"), "event", event)
```
Comment (Jeffwan) [inline ray-operator/controllers/ray/raycluster_controller.go:127]: em. it means we log all other unrelated events which is unnecessary? Let's say there're some other components throw events, it will be printed in kuberay controller log. I feel we should remove this line

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 2]
```diff
@@ -121,27 +123,42 @@ func (r *RayClusterReconciler) eventReconcile(request ctrl.Request, event *corev
...
+	if err := r.List(context.TODO(), &pods, options...); err != nil {
 		return ctrl.Result{}, err
 	}
 
-	for _, item := range pods.Items {
-		if item.Name == event.InvolvedObject.Name {
-			if unhealthyPod != nil {
-				return ctrl.Result{}, fmt.Errorf("duplicate pod found")
+	if len(pods.Items) == 0 {
+		r.Log.Info("no pods found")
```
Comment (Jeffwan) [inline ray-operator/controllers/ray/raycluster_controller.go:146]: Log.Warn - no pods found from event %, event.name

**Review by Jeffwan — COMMENTED** [review]
overall looks good to me

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 3]
```diff
@@ -817,7 +834,33 @@ func (r *RayClusterReconciler) SetupWithManager(mgr ctrl.Manager, reconcileConcu
 			predicate.LabelChangedPredicate{},
 			predicate.AnnotationChangedPredicate{},
 		))).
-		Watches(&source.Kind{Type: &corev1.Event{}}, &handler.EnqueueRequestForObject{}).
+		Watches(&source.Kind{Type: &corev1.Event{}},
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(predicate.Funcs{
+				CreateFunc: func(e event.CreateEvent) bool {
+					// get the v1.Event object
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:841]: nit: The comment above this code snippet appears to be unnecessary as the code is self-explanatory and can be easily understood without additional comments.
```suggestion
```

File: ray-operator/controllers/ray/raycluster_controller_fake_test.go [hunk 4]
```diff
@@ -365,6 +373,149 @@ func tearDown(t *testing.T) {
 	PrioritizeWorkersToDelete = false
 }
 
+func TestReconcile_UnhealthyEvent(t *testing.T) {
+	setupTest(t)
+	defer tearDown(t)
+
+	testPodName := "eventPod"
+	testPodEventName := fmt.Sprintf("%s.15f0c0c5c5c5c5c5", testPodName)
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller_fake_test.go:381]: What is "15f0c0c5c5c5c5c5"?

File: ray-operator/controllers/ray/raycluster_controller_fake_test.go [hunk 5]
```diff
@@ -365,6 +373,149 @@ func tearDown(t *testing.T) {
...
+
+	// add a pod in a different namespace
+	newPods := []runtime.Object{
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller_fake_test.go:384]: The specifications of the Pods are very similar (some parts are the same.). Can we use variables instead of expanding all of them?

File: ray-operator/controllers/ray/raycluster_controller_fake_test.go [hunk 6]
```diff
@@ -365,6 +373,149 @@ func tearDown(t *testing.T) {
 	PrioritizeWorkersToDelete = false
 }
 
+func TestReconcile_UnhealthyEvent(t *testing.T) {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller_fake_test.go:376]: This test add many Pods that may distract attention. We should only focus on unhealthy events here. The following is my suggestion:

* Step 1: Create a new Pod in a new namespace (e.g. name: "new_pod", namespace: "new_namespace").
* Step 2: Create a fake client for the Pod only.
* Step 3: Check whether only 1 Pod in the "new_namespace".
* Step 4: Create an unhealthy event for "new_pod"
* Step 5: Reconcile the event
* Step 6: Check `pod.Annotations[common.RayNodeHealthStateAnnotationKey]`

**Review by kevin85421 — COMMENTED** [review]
Thanks for the contribution! I will check out this branch and test it by myself next week.

File: ray-operator/Makefile [hunk 7]
```diff
@@ -63,7 +63,7 @@ ENVTEST_ASSETS_DIR=$(shell pwd)/testbin
 test: manifests generate fmt vet ## Run tests.
 	mkdir -p ${ENVTEST_ASSETS_DIR}
 	test -f ${ENVTEST_ASSETS_DIR}/setup-envtest.sh || curl -sSLo ${ENVTEST_ASSETS_DIR}/setup-envtest.sh https://raw.githubusercontent.com/kubernetes-sigs/controller-runtime/v0.7.2/hack/setup-envtest.sh
-	export ENVTEST_K8S_VERSION=1.24.2; source ${ENVTEST_ASSETS_DIR}/setup-envtest.sh; fetch_envtest_tools $(ENVTEST_ASSETS_DIR); setup_envtest_env $(ENVTEST_ASSETS_DIR); go test ./... -coverprofile cover.out
+	export ENVTEST_K8S_VERSION=1.24.2; source ${ENVTEST_ASSETS_DIR}/setup-envtest.sh; fetch_envtest_tools $(ENVTEST_ASSETS_DIR); setup_envtest_env $(ENVTEST_ASSETS_DIR); go test ./... -v -coverprofile cover.out
```
[the remark this answers is not in the document]
Comment (wilsonwang371) [reply ray-operator/Makefile:66]: this is for verbose output. it can be helpful in case of failure.

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 8]
```diff
@@ -121,27 +123,42 @@ func (r *RayClusterReconciler) eventReconcile(request ctrl.Request, event *corev
...
-	for _, item := range pods.Items {
-		if item.Name == event.InvolvedObject.Name {
-			if unhealthyPod != nil {
-				return ctrl.Result{}, fmt.Errorf("duplicate pod found")
+	if len(pods.Items) == 0 {
+		r.Log.Error(fmt.Errorf("no pod found for event"), "no pod found for event",
+			"event", event)
+		return ctrl.Result{}, nil
+	} else if len(pods.Items) > 1 {
+		// This happens when we use fake client
```
[the remark this answers is not in the document]
Comment (wilsonwang371) [reply ray-operator/controllers/ray/raycluster_controller.go:148]: this happens during test which we do not have the server side filtering available.

File: ray-operator/controllers/ray/raycluster_controller_fake_test.go (hunk 4 above)
Comment (wilsonwang371) [reply ray-operator/controllers/ray/raycluster_controller_fake_test.go:381]: just a random number for the event.

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 9]
```diff
@@ -121,27 +123,42 @@ func (r *RayClusterReconciler) eventReconcile(request ctrl.Request, event *corev
...
 			}
-			unhealthyPod = &item
 		}
+	} else {
+		unhealthyPod = &pods.Items[0]
 	}
 
-	if unhealthyPod == nil || unhealthyPod.Annotations == nil {
+	if unhealthyPod.Annotations == nil {
 		r.Log.Info("pod not found or no valid annotations", "pod name", event.InvolvedObject.Name)
```
[the remark this answers is not in the document]
Comment (wilsonwang371) [reply ray-operator/controllers/ray/raycluster_controller.go:161]: i didnt get it.

File: ray-operator/controllers/ray/raycluster_controller_fake_test.go (hunk 5 above)
Comment (wilsonwang371) [reply ray-operator/controllers/ray/raycluster_controller_fake_test.go:384]: This is a test case. In this test case, we have `common.RayNodeLabelKey: "yes",` specified. I know some parts are the same but putting them in a separate place might not be good for test cases.

File: ray-operator/controllers/ray/raycluster_controller_fake_test.go (hunk 6 above)
Comment (wilsonwang371) [reply ray-operator/controllers/ray/raycluster_controller_fake_test.go:376]: The reason I am having more than one pods is that I was testing the filter. This test helped me to realized that some filters will only work on the server side. So I think this many pods can later be helpful if we want to make this test more robust.

File: ray-operator/controllers/ray/raycluster_controller_fake_test.go (hunk 4 above)
Comment (kevin85421) [reply ray-operator/controllers/ray/raycluster_controller_fake_test.go:381]: Can we add a comment for it? Others may feel confusing for the random string.

File: ray-operator/controllers/ray/raycluster_controller_fake_test.go (hunk 5 above)
Comment (kevin85421) [reply ray-operator/controllers/ray/raycluster_controller_fake_test.go:384]: Let me explain it more. My meaning is that:

```go
spec :=  corev1.PodSpec{
  Containers: []corev1.Container{
	  {
		  Name:    "ray-worker",
		  Image:   "rayproject/ray:2.2.0",
		  Command: []string{"echo"},
		  Args:    []string{"Hello Ray"},
	  },
  },
}

&corev1.Pod {
   ...
   Spec: spec (Maybe need to use Deep copy)
}

&corev1.Pod {
   ...
   Spec: spec (Maybe need to use Deep copy)
}

```

instead of 

```go
&corev1.Pod {
   ...
   Spec: corev1.PodSpec{
    Containers: []corev1.Container{
	    {
		    Name:    "ray-worker",
		    Image:   "rayproject/ray:2.2.0",
		    Command: []string{"echo"},
		    Args:    []string{"Hello Ray"},
	    },
    },
  }
}

&corev1.Pod {
   ...
   Spec: corev1.PodSpec{
    Containers: []corev1.Container{
	    {
		    Name:    "ray-worker",
		    Image:   "rayproject/ray:2.2.0",
		    Command: []string{"echo"},
		    Args:    []string{"Hello Ray"},
	    },
    },
  }
}

```

File: ray-operator/controllers/ray/raycluster_controller.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (kevin85421) [reply ray-operator/controllers/ray/raycluster_controller.go:148]: Would you mind explaining it more? In my understanding, it is impossible to have two Pods with the same name in an actual situation? Why do we need to test this case? Do I miss anything? Thanks!

File: ray-operator/controllers/ray/raycluster_controller.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (kevin85421) [reply ray-operator/controllers/ray/raycluster_controller.go:161]: If Pod is not found, this function will be returned by the following logic:

```go
if len(pods.Items) == 0 {
  r.Log.Error(fmt.Errorf("no pod found for event"), "no pod found for event",
	  "event", event)
  return ctrl.Result{}, nil
}
```

Hence, the log message should update from `"pod not found or no valid annotations"` to `"no unhealthy annotations on the unhealthy Pod"`.

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 10]
```diff
@@ -121,27 +123,41 @@ func (r *RayClusterReconciler) eventReconcile(request ctrl.Request, event *corev
 	// we only care about pod events
 	if event.InvolvedObject.Kind != "Pod" || event.Type != "Warning" || event.Reason != "Unhealthy" ||
 		!strings.Contains(event.Message, "Readiness probe failed") {
+		// This is not supposed to happen
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:126]: Should we surface the issue here rather than `return` without doing anything?

File: ray-operator/controllers/ray/raycluster_controller_fake_test.go (hunk 6 above)
Comment (kevin85421) [reply ray-operator/controllers/ray/raycluster_controller_fake_test.go:376]: > This test helped me to realized that some filters will only work on the server side.

Which filters? What's the difference between the behavior in our tests and actual k8s cluster? If the behaviors are different, how do we test the functionality? Thanks!

Comment (kevin85421) [reply ray-operator/controllers/ray/raycluster_controller_fake_test.go:376]: Would you mind adding some comments for the test? Thanks!

File: ray-operator/controllers/ray/raycluster_controller_fake_test.go (hunk 5 above)
Comment (kevin85421) [reply ray-operator/controllers/ray/raycluster_controller_fake_test.go:390]: There are a lot of technical debts in KubeRay now. It's good for us to reduce the system's complexity in every PR : )

File: ray-operator/controllers/ray/raycluster_controller.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (wilsonwang371) [reply ray-operator/controllers/ray/raycluster_controller.go:158]: This is happening during test in which we are using fakeclient which does not support these different kinds of filters applied in the ListOption. If we don't have this, the test will fail which is not because of a real failure.

**Comment by Jeffwan** [comment]
@wilsonwang371 Please help address @kevin85421's comment.

**Comment by kevin85421** [comment]
Hi @wilsonwang371, would you mind rebasing with the master's branch? Because of the tight deadline for the release v0.5.0, @architkulkarni and I will merge this PR today and file a new PR to fix/refactor if we find any bugs in this PR. Thanks!

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 11]
```diff
@@ -121,28 +133,43 @@ func (r *RayClusterReconciler) eventReconcile(request ctrl.Request, event *corev
...
 	if event.InvolvedObject.Kind != "Pod" || event.Type != "Warning" || event.Reason != "Unhealthy" ||
 		!strings.Contains(event.Message, "Readiness probe failed") {
+		// This is not supposed to happen since we already filter events in the watch
+		msg := fmt.Sprintf("unexpected event, we should have already filtered these conditions: %v", event)
+		r.Log.Error(fmt.Errorf(msg), msg, "event", event)
 		return ctrl.Result{}, nil
 	}
 
 	_ = r.Log.WithValues("event", request.NamespacedName)
-	r.Log.Info("reconcile RayCluster Event", "event name", request.Name)
```
Comment (architkulkarni) [inline ray-operator/controllers/ray/raycluster_controller.go:128]: Why remove this log? Is it no longer useful?

File: ray-operator/controllers/ray/raycluster_controller_fake_test.go [hunk 12]
```diff
@@ -365,6 +373,158 @@ func tearDown(t *testing.T) {
...
+	assert.Equal(t, len(testPods)-1, len(podList.Items), "Init pod list len is wrong")
+
+	testRayClusterReconciler := &RayClusterReconciler{
+		Client:   fakeClient,
+		Recorder: &record.FakeRecorder{},
+		Scheme:   scheme.Scheme,
+		Log:      ctrl.Log.WithName("controllers").WithName("RayCluster"),
+	}
+
+	// add event for reconcile
```
Comment (architkulkarni) [inline ray-operator/controllers/ray/raycluster_controller_fake_test.go:487]: Does it make sense to also add an event which should get filtered out and not trigger a reconcile?

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 13]
```diff
@@ -814,7 +841,31 @@ func (r *RayClusterReconciler) SetupWithManager(mgr ctrl.Manager, reconcileConcu
...
 			predicate.AnnotationChangedPredicate{},
 		))).
-		Watches(&source.Kind{Type: &corev1.Event{}}, &handler.EnqueueRequestForObject{}).
+		Watches(&source.Kind{Type: &corev1.Event{}},
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(predicate.Funcs{
+				CreateFunc: func(e event.CreateEvent) bool {
+					if eventObj, ok := e.Object.(*corev1.Event); ok {
+						if eventObj.InvolvedObject.Kind != "Pod" || eventObj.Type != "Warning" ||
+							eventObj.Reason != "Unhealthy" || !strings.Contains(eventObj.Message, "Readiness probe failed") {
```
Comment (architkulkarni) [inline ray-operator/controllers/ray/raycluster_controller.go:850]: Is there a more programmatic way to get the cause of the unhealthiness, or is scanning the message string standard practice?  Also, for my education, why do we only care about failures in the readiness probe and not other pod unhealthy events?  Maybe @kevin85421 you can also respond if you know the answer

**Review by architkulkarni — APPROVED** [review]
Overall looks good to me

**Comment by architkulkarni** [comment]
@wilsonwang371 Do you mind adding a bit more detail to the PR description?

**Comment by wilsonwang371** [comment]
Sorry, I am a little bit busy these days. I just took some time to rebase and update the comments. Let's wait and see how's @jeevb reply on this latest patch.

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 14]
```diff
@@ -43,10 +46,20 @@ var (
...
 	EnableBatchScheduler      bool
+
+	// Definition of a index field for pod name
+	podUIDIndexField = "metadata.uid"
 )
 
 // NewReconciler returns a new reconcile.Reconciler
 func NewReconciler(mgr manager.Manager) *RayClusterReconciler {
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &corev1.Pod{}, podUIDIndexField, func(rawObj client.Object) []string {
+		pod := rawObj.(*corev1.Pod)
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:57]: The following change may fix https://github.com/ray-project/kuberay/issues/601#issuecomment-1486149187. cc @jeevb 

```
pod := rawObj.(*corev1.Pod)
owner := metav1.GetControllerOf(pod)
if owner == nil || owner.APIVersion != rayiov1alpha1.GroupVersion.String() || owner.Kind != "RayCluster" {
	return nil
}
return []string{string(pod.UID)}
```

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 15]
```diff
@@ -120,28 +133,43 @@ func (r *RayClusterReconciler) eventReconcile(request ctrl.Request, event *corev
...
 
 	_ = r.Log.WithValues("event", request.NamespacedName)
-	r.Log.Info("reconcile RayCluster Event", "event name", request.Name)
 
-	if err := r.List(context.TODO(), &pods,
-		client.InNamespace(event.InvolvedObject.Namespace)); err != nil {
+	options := []client.ListOption{
+		client.MatchingFields(map[string]string{podUIDIndexField: string(event.InvolvedObject.UID)}),
+		client.InNamespace(event.InvolvedObject.Namespace),
+		client.MatchingLabels(map[string]string{common.RayNodeLabelKey: "yes"}),
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:147]: Remove all `common.RayNodeLabelKey`

**Review by kevin85421 — COMMENTED** [review]
If you do not have time to update & test this PR tomorrow, I can merge this PR and open a new patch to fix the issues.

File: ray-operator/controllers/ray/raycluster_controller.go (hunk 13 above)
Comment (wilsonwang371) [reply ray-operator/controllers/ray/raycluster_controller.go:851]: readiness probe failure will trigger liveness probe failure. readiness probe will not automatically trigger a pod rebuild but liveness probe failure will.

File: ray-operator/controllers/ray/raycluster_controller.go (hunk 15 above)
Comment (wilsonwang371) [reply ray-operator/controllers/ray/raycluster_controller.go:147]: If you have time to help this, that will be good.

**Comment by wilsonwang371** [comment]
> If you do not have time to update & test this PR tomorrow, I can merge this PR and open a new patch to fix the issues.

If you have time to help with this, that will be good.

File: ray-operator/controllers/ray/raycluster_controller.go (hunk 15 above)
Comment (wilsonwang371) [reply ray-operator/controllers/ray/raycluster_controller.go:147]: why?

**Review by Jeffwan — APPROVED** [review]

**Comment by kevin85421** [comment]
Merge this PR, and will file a separate PR to fix and improve some minor issues.

File: ray-operator/controllers/ray/raycluster_controller.go (hunk 15 above)
Comment (kevin85421) [reply ray-operator/controllers/ray/raycluster_controller.go:147]: I found out that I misunderstood some things. Initially, I thought that we should implement all filter logic in the resource event handler (i.e. `CreateFunc`) and return true only when an event is issued by a Ray Pod. With this approach, we can make the function `eventReconcile` only handle event processing and not worry about event filtering. Therefore, the label `common.RayNodeLabelKey` is not necessary.

However, I have since realized that the owners of the unhealthy events are Pods, which may include Ray Pods and other Pods. Therefore, we cannot use `GetControllerOf` to get the owner and determine whether the event belongs to a Ray Pod or not by checking whether the owner's kind is "RayCluster" or not.

Hence, the implementation will be more complex. If we want to distinguish whether an event belongs to a Ray Pod or not, we need to

* Get the UID of the event's InvolvedObject.
* Use the UID to look up Pods from the local cache.
* Use `GetControllerOf(pod)` to get the owner, and check whether the `kind` is `RayCluster` or not.

I will make the function `eventReconcile` easier to understand, but we need to have more I/O (read local cache) which may cause performance overhead. 

In summary, we should use `common.RayNodeLabelKey` to distinguish Ray Pods in `eventReconcile` until we understand the cost of the look-up operation.

File: ray-operator/controllers/ray/raycluster_controller.go (hunk 14 above)
Comment (kevin85421) [reply ray-operator/controllers/ray/raycluster_controller.go:57]: ^ 

Above is wrong. This is used to enable efficient lookup operations for the local cache. The informer will still fetch cluster status from Kubernetes APIServer. That is, events which are unrelated to Ray Pods will finally be fetched by the informer. Sorry for the misleading.
